### PR TITLE
[JENKINS-30269] Add pipeline support for resource locking

### PR DESCRIPTION
**Pull request by amuniz**
[JENKINS-30269](https://issues.jenkins-ci.org/browse/JENKINS-30269)

This PR allows to use this plugin in a Pipeline job.

```groovy
echo 'Starting'
lock('my-resource-name') {
  echo 'Do something here that requires unique access to the resource'
  // any other build will wait until the one locking the resource leaves this block
}
echo 'Finish'
```

Lockable resources are defined as Jenkins global configuration and they are shared with any other build in Jenkins using this plugin (Pipeline or not), so mutual exclusion works between Freestyle builds and Pipeline builds for example.

@reviewbybees

### CODE REVIEW

File: src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+			body.cancel(cause);
+		}
+		if (task != null) {
+			task.cancel(false);
+		}
+		getContext().onFailure(cause);
+	}
+
+	/**
+	 * Records that a build was canceled because it reached a milestone but a
```
Comment (recampbell) [inline src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java]: This looks like a cut & paste description. It seems like this is a build being canceled because the maxWaiting parameter has been exceeded.

Perhaps this should be renamed to WaitLimitExceededCause?

File: src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+		}
+
+		public Run<?, ?> getNewerBuild() {
+			return Run.fromExternalizableId(newerBuild);
+		}
+
+		@Override
+		public String getShortDescription() {
+			if (newerBuild != null) {
+				return "Superseded by " + getNewerBuild().getDisplayName();
```
Comment (recampbell) [inline src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java]: It would be nice to know that this was build was superseded because the ${%Limit waiting builds} field was exceeded. Should also include the name of the resource which was locked. Otherwise, I have no idea why a build was canceled?

File: pom.xml [hunk 3]
```diff
@@ -54,6 +70,48 @@
...
 			<version>1.5</version>
 		</dependency>
+		<dependency>
+			<groupId>org.jenkins-ci.plugins.workflow</groupId>
+			<artifactId>workflow-step-api</artifactId>
+			<version>${workflow.version}</version>
+		</dependency>
+		<dependency>
+			<groupId>org.jenkins-ci.plugins</groupId>
+			<artifactId>matrix-project</artifactId>
```
Comment (jglick) [inline pom.xml]: Yuck. Wonder if this can be avoided somehow.

File: pom.xml [hunk 4]
```diff
@@ -54,6 +70,48 @@
...
+			<artifactId>annotations</artifactId>
+			<version>3.0.0</version>
+		</dependency>
+
+		<!-- Testing scope -->
+		<dependency>
+			<groupId>com.infradna.tool</groupId>
+			<artifactId>bridge-method-annotation</artifactId>
+			<version>1.14</version>
+			<optional>true</optional>
```
Comment (jglick) [inline pom.xml]: Well, this is not `test` scope.

File: src/main/resources/org/jenkins/plugins/lockableresources/LockStep/help-maxWaiting.html [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<div>
+	<p>
+		By default builds a queued without any restriction on waiting to lock a common resource
+		Setting this parameter a limit is forced in the number of waiting builds. Any build trying to lock after
+		the limit is reached will trigger a cancel round trip which will abort the older waiting build.
+	</p>
+	<p>
+		For example, if builds #3 and #4 are waiting to lock the resource and build #5 reaches the lock step,
+		then build #3 will be cancelled in favor of #5 which is newer.
```
Comment (jglick) [inline src/main/resources/org/jenkins/plugins/lockableresources/LockStep/help-maxWaiting.html]: Do we actually need this feature? To my mind the `milestone` step is supposed to be taking care of that kind of ordering. In other words, 3, 4, and 5 would all wait for the lock. You _can_ take build number into consideration when deciding who gets the lock when it is released, giving preference to 5; then when 5 reaches its next milestone, 3 and 4 would get canceled automatically.

Comment (jglick) [inline src/main/resources/org/jenkins/plugins/lockableresources/LockStep/help-maxWaiting.html]: Thus `CanceledCause` and associated machinery could also be deleted. `MilestoneStepExecution` would take care of all this.

File: src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+		// the body was already started, nothing to do here
+	}
+
+	private void tryLock(long delay) {
+		task = Timer.get().schedule(new Runnable() {
+			@Override
+			public void run() {
+				task = null;
+				if (!lockAndProceed()) {
+					retry(id, 1000); // try to lock every second
```
Comment (jglick) [inline src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java]: Does not make sense here. When a lock is released, there should be some sort of notification inside this plugin, and the paused step should be unblocked. There is no need to “poll” for a lock defined inside Jenkins itself.

Comment (jglick) [inline src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java]: You just need to save the `StepContext`s of waiting threads associated with a given lock.

File: src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+				}
+			}
+		}, delay, TimeUnit.MILLISECONDS);
+	}
+
+	private synchronized boolean lockAndProceed() {
+		LockableResourcesStruct resourceHolder = new LockableResourcesStruct(step.resource);
+		LOGGER.finest("Trying to acquire [" + step.resource + "] by " + run.getExternalizableId());
+		// Apply maxWaiting if needed
+		Run<?, ?> older = LockableResourcesManager.get().addToQueue(resourceHolder.required, run, step.maxWaiting);
```
Comment (jglick) [inline src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java]: I think the whole storage system here is wrong. For Pipeline builds, it does not suffice to save the build reference. You need to know the `StepContext`. Consider a Jenkins instance with a single (Pipeline) job which runs one build:

```groovy
parallel a: {
  sleep 5
  lock('stuff') {
    sleep 5
  }
}, b: {
  lock('stuff') {
    sleep 10
  }
}
```

Thread `b` should acquire the lock, then five seconds later `a` starts to wait for it, then five seconds later `b` releases it and exits, allowing `a` to lock it, then five seconds later `a` releases the lock and the build finishes. The handoff from `b` to `a` should happen by the callback from `b`’s block looking up a waiting `StepContext`, in this case `a`’s, and instructing it to proceed via `newBodyInvoker`.

**Comment by amuniz** [comment]
@recampbell thanks for your review, I believe all your comments were fixed.

@jglick I switched to a non-polling design. I think it's much better now. Thanks for your tips!

**Comment by doctori** [comment]
The end of the lock phase seems to generate NPE : 
```
java.lang.NullPointerException
	at org.jenkins.plugins.lockableresources.LockableResource.getNextQueuedContext(LockableResource.java:302)
	at org.jenkins.plugins.lockableresources.LockableResourcesManager.unlock(LockableResourcesManager.java:261)
	at org.jenkins.plugins.lockableresources.LockStepExecution$Callback.unlock(LockStepExecution.java:82)
	at org.jenkins.plugins.lockableresources.LockStepExecution$Callback.finished(LockStepExecution.java:75)
	at org.jenkinsci.plugins.workflow.steps.BodyExecutionCallback$TailCall.onSuccess(BodyExecutionCallback.java:114)
	at org.jenkinsci.plugins.workflow.cps.CpsBodyExecution$SuccessAdapter.receive(CpsBodyExecution.java:311)
	at com.cloudbees.groovy.cps.Outcome.resumeFrom(Outcome.java:73)
	at com.cloudbees.groovy.cps.Continuable.run0(Continuable.java:146)
	at org.jenkinsci.plugins.workflow.cps.CpsThread.runNextChunk(CpsThread.java:164)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.run(CpsThreadGroup.java:277)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.access$000(CpsThreadGroup.java:77)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:186)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:184)
	at org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$2.call(CpsVmExecutorService.java:47)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:112)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

With this kind of configuration :   
```
lock('Integration') {
    node('linux'){
        stage "test"
        sh "echo 'test'"
        sleep time: 1, unit: 'MINUTES'
        sh "echo 'end test'"
    }
}
```

**Comment by amuniz** [comment]
@doctori I guess you were testing after upgrading from a previous version right? I fixed that [here](https://github.com/jenkinsci/lockable-resources-plugin/pull/25/commits/c0e357dc5d28baf7df8c3c33296d09ce244499c4), can you confirm that the NPE is not there now? Thanks.

**Comment by doctori** [comment]
@amuniz It seems to work nicely now. Thank's !

File: src/main/resources/org/jenkins/plugins/lockableresources/LockStep/help-inversePrecedence.html [hunk 8]
```diff
@@ -0,0 +1,8 @@
+<div>
+	<p>
+		By default waiting builds get the lock in the same order the requested to acquire it.
```
Comment (jglick) [inline src/main/resources/org/jenkins/plugins/lockableresources/LockStep/help-inversePrecedence.html]: typo

File: src/main/resources/org/jenkins/plugins/lockableresources/LockStep/config.jelly [hunk 9]
```diff
@@ -0,0 +1,9 @@
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define"
+	xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form">
+	<f:entry title="${%Resource}" field="resource">
+		<f:textbox/>
+	</f:entry>
+	<f:entry title="${%Inverse precedence}" field="inversePrecedence">
```
Comment (jglick) [inline src/main/resources/org/jenkins/plugins/lockableresources/LockStep/config.jelly]: I wonder if this should be checked by default. It seems like the behavior you would normally want.

File: src/main/resources/org/jenkins/plugins/lockableresources/LockStep/help-inversePrecedence.html [hunk 10]
```diff
@@ -0,0 +1,8 @@
...
+	</p>
+	<p>
+		By checking this option the newest build in the waiting queue will get the lock first.
```
Comment (jglick) [inline src/main/resources/org/jenkins/plugins/lockableresources/LockStep/help-inversePrecedence.html]: What if there are builds of multiple jobs waiting? Including multiple builds of a single job?

File: src/main/resources/org/jenkins/plugins/lockableresources/LockStep/help-resource.html [hunk 11]
```diff
@@ -0,0 +1,5 @@
+<div>
+	<p>
+		The resource name to lock as defined in Global settings.
```
Comment (jglick) [inline src/main/resources/org/jenkins/plugins/lockableresources/LockStep/help-resource.html]: Hmm, I guess this means you cannot implicitly define a resource just by starting to lock it? Goes against the grain of Pipeline-as-code.

File: src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+			this.inversePrecedence = inversePrecedence;
+		}
+
+		protected void finished(StepContext context) throws Exception {
+			unlock(context);
+		}
+
+		private void unlock(StepContext context) throws IOException, InterruptedException {
+			if (run == null && buildExternalizableId != null) {
+				run = Run.fromExternalizableId(buildExternalizableId);
```
Comment (jglick) [inline src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java]: Could you not just use `context.get(Run.class)` and omit the `run` and `buildExternalizableId` fields?

File: src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+import org.jenkinsci.plugins.workflow.steps.BodyExecutionCallback;
+import org.jenkinsci.plugins.workflow.steps.StepContext;
+import org.jenkinsci.plugins.workflow.steps.StepContextParameter;
+
+import com.google.inject.Inject;
+
+import hudson.model.Run;
+import hudson.model.TaskListener;
+
+public class LockStepExecution extends AbstractStepExecutionImpl {
```
Comment (jglick) [inline src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java]: missing `serialVersionUID` AFAICT

File: src/main/java/org/jenkins/plugins/lockableresources/LockableResource.java [hunk 14]
```diff
@@ -216,8 +296,44 @@ private void validateQueuingTimeout() {
...
+				int newestIndex = 0;
+				for (StepContext c : queuedContexts) {
+					try {
+						Run<?, ?> run = c.get(Run.class);
+						if (run.getNumber() > newest) {
+							newest = run.getNumber();
+							newestIndex = index;
+						}
+					} catch (Exception e) {
+						// skip this one
```
Comment (jglick) [inline src/main/java/org/jenkins/plugins/lockableresources/LockableResource.java]: At least log the exception.

File: src/main/java/org/jenkins/plugins/lockableresources/LockableResource.java [hunk 15]
```diff
@@ -216,8 +296,44 @@ private void validateQueuingTimeout() {
...
+			if (!inversePrecedence) {
+				return queuedContexts.remove(0);
+			} else {
+				int newest = 0;
+				int index = 0;
+				int newestIndex = 0;
+				for (StepContext c : queuedContexts) {
+					try {
+						Run<?, ?> run = c.get(Run.class);
+						if (run.getNumber() > newest) {
```
Comment (jglick) [inline src/main/java/org/jenkins/plugins/lockableresources/LockableResource.java]: What if `run` is from a different `Job` than the one represented by `newest`? Then the build number comparison will be meaningles.

File: src/main/java/org/jenkins/plugins/lockableresources/LockStep.java [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+
+	public boolean inversePrecedence = false;
+
+	@DataBoundConstructor
+	public LockStep(String resource) {
+		if (resource == null || resource.isEmpty()) {
+			throw new IllegalArgumentException("must specify resource");
+		}
+		if (LockableResourcesManager.get().fromName(resource) == null) {
+			throw new IllegalArgumentException("resource [" + resource + "] does not exist, missing global definition?");
```
Comment (jglick) [inline src/main/java/org/jenkins/plugins/lockableresources/LockStep.java]: As I mentioned in the help file, I wonder if it would be appropriate to automatically create resources in the step, at least with a `create: true` option or the like, so that scripts can be self-contained.

File: src/test/java/org/jenkins/plugins/lockableresources/LockStepTest.java [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+				p.setDefinition(new CpsFlowDefinition(
+						"lock('resource1') {\n" +
+						"	echo 'Locked'\n" +
+						"}\n" +
+						"echo 'Finish'"
+				));
+
+				FreeStyleProject f = story.j.createFreeStyleProject("f");
+				f.addProperty(new RequiredResourcesProperty("resource1", null, null, null));
+				f.getBuildersList().add(new Shell("sleep 10"));
```
Comment (jglick) [inline src/test/java/org/jenkins/plugins/lockableresources/LockStepTest.java]: You can use a `MockBuilder` (or `TestBuilder`?) which waits for a `Semaphore`. More reliable, portable, and faster.

File: src/test/java/org/jenkins/plugins/lockableresources/LockStepTest.java [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+				f.scheduleBuild2(0).waitForStart();
+
+				WorkflowRun b1 = p.scheduleBuild2(0).waitForStart();
+				story.j.waitForMessage("[resource1] is locked, waiting...", b1);
+
+				// Wait for lock after the freestyle finishes
```
Comment (jglick) [inline src/test/java/org/jenkins/plugins/lockableresources/LockStepTest.java]: Might be interested to test the scenario that

* a Pipeline build starts
* it locks a resource
* a freestyle job is scheduled, but waits in queue
* Jenkins restarts
* the Pipeline build releases the lock and exits
* now the freestyle build is sent to an executor

which proves that the queue logic works properly with serialized lock state.

**Comment by jglick** [comment]
I noted some minor bugs and suggestions, but overall this is looking good. :bee:

File: src/main/resources/org/jenkins/plugins/lockableresources/LockStep/help-resource.html (hunk 11 above)
Comment (amuniz) [inline src/main/resources/org/jenkins/plugins/lockableresources/LockStep/help-resource.html]: Yes, but it's needed to keep interoperability with freestyle (otherwise the whole design of this plugin would need to be revisited). This case is similar to `tool` step.

File: src/test/java/org/jenkins/plugins/lockableresources/LockStepTest.java (hunk 18 above)
Comment (amuniz) [inline src/test/java/org/jenkins/plugins/lockableresources/LockStepTest.java]: Sure. I'll do.

File: src/main/resources/org/jenkins/plugins/lockableresources/LockStep/help-inversePrecedence.html (hunk 10 above)
Comment (amuniz) [inline src/main/resources/org/jenkins/plugins/lockableresources/LockStep/help-inversePrecedence.html]: The build number is being used currently to pick up the newest, so it works for multiple builds of the same job but not for other jobs... need to fix it (probably using `run.getStartTimeInMillis()`). Thanks for finding it!

File: src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java (hunk 12 above)
Comment (amuniz) [inline src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java]: I suppose, will try it.

File: pom.xml (hunk 3 above)
Comment (amuniz) [inline pom.xml]: I don't think so, it's a hard dependency (imports from `matrix-project` in the code mainly to skip lock in defined cases).

File: src/main/resources/org/jenkins/plugins/lockableresources/LockStep/config.jelly (hunk 9 above)
Comment (amuniz) [inline src/main/resources/org/jenkins/plugins/lockableresources/LockStep/config.jelly]: No strong opinion. Some days ago I was asked by a Jenkins user in Seville JAM about how to change the behavior of `stage` (newer builds first), they just want to execute all builds as they come.

File: src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java (hunk 13 above)
Comment (amuniz) [inline src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java]: Fixed.

File: src/main/java/org/jenkins/plugins/lockableresources/LockStep.java (hunk 16 above)
Comment (amuniz) [inline src/main/java/org/jenkins/plugins/lockableresources/LockStep.java]: Already answered in the other comment. Creating resources as the build goes does not make much sense, as a new resource would be created on each build, unless we somehow uniquely identify a `lock` step and link it to a generated resource, which seems overkill to me.

Moreover, Freestyle interoperability would require to create resources in global configuration.

This case is similar to `tool`, which requires the tool to be defined globally first.

**Comment by amuniz** [comment]
@jglick a final review please?

File: src/main/java/org/jenkins/plugins/lockableresources/LockStep.java (hunk 16 above)
Comment (jglick) [inline src/main/java/org/jenkins/plugins/lockableresources/LockStep.java]: > unless we somehow uniquely identify a `lock` step

Huh? The step is already specifying a resource name. My suggestion was merely to create such a resource if it does not already exist.

```java
LockableResourcesManager mgr = LockableResourcesManager.get();
if (mgr.fromName(resource) == null) {
    mgr.getResources().add(new LockableResource(resource));
    mgr.save();
}
```

Something like that.

> Freestyle interoperability would require to create resources in global configuration.

Sure, but for a Pipeline-only installation, you would not need this step. I would not expect most users of this plugin to be mixing freestyle and Pipeline builds on the same resource; it is probably a corner case.

> This case is similar to `tool`

Not really. In the case of a `ToolInstallation`, there is real configuration you are providing, like a `ToolInstaller`. The name of the tool is merely an identifier for that configuration. A resource in the common case is just an opaque token. `LockableResource` can apparently take a description and some labels, but these seem quite optional.

**Comment by jglick** [comment]
Lack of lock resource autocreation still means that this is not a viable substitute for `StageStep.concurrency` in the context of Pipeline-as-Code, because an administrator would need to intervene to define resources for each project using the functionality.

**Comment by amuniz** [comment]
> Lack of lock resource autocreation still means that this is not a viable substitute for StageStep.concurrency

Yes, agree. I'll add an option to auto-create the resource if it does not exist (active by default).

File: src/test/java/org/jenkins/plugins/lockableresources/LockStepTest.java [hunk 19]
```diff
@@ -0,0 +1,305 @@
...
+
+				story.j.waitForMessage("acquired lock on [resource1]", fb1);
+			}
+		});
+	}
+
+
+	private void defineResource(String r) {
+		LockableResourcesManager.get().getResources().add(new LockableResource(r));
+		LockableResourcesManager.get().save();
```
Comment (jglick) [inline src/test/java/org/jenkins/plugins/lockableresources/LockStepTest.java]: Could be replaced with `createResource` (and inlined).

File: src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+			// It's granted to contain one item (and only one for now)
+			this.resourceHolder = resourceHolder;
+			this.inversePrecedence = inversePrecedence;
+		}
+
+		protected void finished(StepContext context) throws Exception {
+			unlock(context);
+		}
+
+		private void unlock(StepContext context) throws IOException, InterruptedException {
```
Comment (jglick) [inline src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java]: Could be inlined into `finish`.

File: src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+			LOGGER.finest("Lock released on [" + resourceHolder.required.get(0) + "]");
+		}
+
+		private static final long serialVersionUID = 1L;
+
+	}
+
+	@Override
+	public void stop(Throwable cause) throws Exception {
+		// NO-OP
```
Comment (jglick) [inline src/main/java/org/jenkins/plugins/lockableresources/LockStepExecution.java]: :bug: this will be called if the build is interrupted while we are waiting for the lock. It should abort the wait and clean up (such as by removing this context from the list of waiters).

File: src/main/resources/org/jenkins/plugins/lockableresources/LockStep/help-resource.html (hunk 11 above)
Comment (jglick) [inline src/main/resources/org/jenkins/plugins/lockableresources/LockStep/help-resource.html]: (resolved)

**Comment by jglick** [comment]
:bee:

File: pom.xml [hunk 22]
```diff
@@ -13,7 +13,7 @@
 	<parent>
 		<groupId>org.jenkins-ci.plugins</groupId>
 		<artifactId>plugin</artifactId>
-		<version>1.509.4</version>
+		<version>1.609.1</version>
```
Comment (kwhetstone) [inline pom.xml]: Would this be a good time to change the parent to "plugin-pom"?  It's optional; you'd gain some dependencies like annotations & can specify your desired Jenkins version.

**Comment by kwhetstone** [comment]
:bee:

File: pom.xml (hunk 22 above)
Comment (amuniz) [inline pom.xml]: Yeah, but upgrading to the new plugins parent POM usually requires more fixed here and there. Let me do that a separate PR.

Comment (kwhetstone) [inline pom.xml]: Sounds good; sometimes getting through the test upgrade takes a bit more time.
